src/metabase/pulse/render/body.clj

```diff
@@ -1030,16 +1030,31 @@
       [:img {:style (style/style {:display :block :width :100%})
              :src   (:image-src image-bundle)}]]}))
 
+(defn- all-unique?
+  [funnel-rows]
+  (let [ks (into #{} (map :key funnel-rows))]
+    (= (count ks) (count funnel-rows))))
+
+(defn- build-funnel-rows
+  [funnel-rows raw-rows]
+  (let [funnel-rows (map-indexed (fn [idx row] (assoc row :idx (inc idx))) funnel-rows)
```

Comment:
I usually don't like shadowing bindings (and would have called this `indexed-funnel-rows` myself) but I guess it's not a huge deal in this particular fn

Reply:
I used to belong to the "never shadow" camp. I had my mind changed after working in Rust for a while though, where it's fairly idiomatic to use shadowing when transforming values. The language has a pretty strict linter, and the linter for shadowing is that it's only an error if the expression for the new binding doesn't use the old binding.

This is partly a stylistic thing, to make functional interfaces look more imperative (😫), but also practical as it moves the value (often off limits through the borrow checker) out of scope. When the conceptual entity is the same, it's nice to avoid a proliferation of similar or awkward names, and rules out the failure mode where you can pick the wrong one.

That said, `indexed-funnel-rows` is a very natural name, so I can go either way on this case. Mostly mentioning because we have a fair amount of code that uses this pattern of shadowing, and I don't think we need a holy war against it 😄